File: sidecar/agent.py
"""The ADK agent — one loop that can call tools repeatedly before answering.

The problem this solves: routing used to give the model exactly one tool call
per turn and then stop. When a question needed something the model didn't have
— the next page of results, what a thread actually says — it couldn't ask, so
it invented. Every hallucination this project hit came from that gap.

Here the model calls a tool, sees the result, and decides again. ADK's Runner
drives that loop; the tools below are plain Python functions, and ADK builds
their schemas from the signature and docstring. Nothing hand-written.
"""
import os
from typing import Any, Dict, Optional

from google.adk.agents import LlmAgent
from google.adk.models.lite_llm import LiteLlm
from google.adk.runners import Runner
from google.adk.sessions import DatabaseSessionService
from google.genai import types

import corpus
import search
from llm import LLAMA_SERVER_URL

APP_NAME = "techne"

# One user per machine — this is a desktop app, not a service. ADK requires a
# user_id, so it is a constant; conversations, not users, are the dimension
# that varies.
USER_ID = "local"

# Beside the corpus cache. Contents/Resources is read-only in a signed .app, so
# anything written at runtime lives under Application Support.
DB_DIR = os.path.expanduser("~/Library/Application Support/com.technesystems.techne")
DB_PATH = os.path.join(DB_DIR, "sessions.db")

BASE_INSTRUCTION = (
    "You help a user explore Hacker News. Use search_threads to find "
    "discussions on a topic, and get_thread to read one in detail when the "
    "user asks about a specific result. Answer from what the tools return — do "
    "not invent discussions, titles, or links. If you do not have enough "
    "information to answer, say so plainly rather than guessing. Do not show "
    "thread ids to the user; they are for your own tool calls."
    "\n\nA message may begin with a note about a discussion the user has open "
    "in front of them. When they say \"this thread\", \"this discussion\", "
    "\"this\", or \"this one\", they mean that one — not an item from a list "
    "of search results. Answer from the note; there is no need to search."
)

# The thread the user has dragged into the chat travels WITH the message it was
# sent alongside, rather than living in session state.
#
# State was the wrong home for it, and failed in a way worth recording. State
# feeds the system prompt, which has no position in time — as far as the model
# can tell it has always said this — while a thread just read with get_thread
# sits in the newest message. Asked to explain "this thread", the model picked
# the recent one. That is a correct reading of what it could see, and no amount
# of instruction wording outranked it (measured: it still chose the wrong thread
# with an explicit "the user opened it just now" line in the prompt).
#
# An attachment belongs to the turn it was attached to. Putting it there makes
# the pin the most recent thing in the conversation, which is what the user
# means by "this", and it makes history self-describing: an old turn keeps its
# attachment, so unpinning cannot make the model disown an earlier answer as
# invented — which is what deleting the state used to cause.
def _with_attachment(message: str, thread: Optional[Dict[str, Any]]) -> str:
    if not thread:
        return message
    return (
        "[The user has this discussion open in front of them:\n"
        f"\"{thread.get('story_title', '')}\" — {thread.get('theme', '')}.\n"
        f"What people said: {thread.get('summary', '')}]\n\n"
        f"{message}"
    )


def search_threads(keyword_filter: str) -> Dict[str, Any]:
    """Search Hacker News discussion threads by topic.

    Args:
        keyword_filter: The topic or keywords to search discussions for.
    """
    # ADK reads this function's name, type hints and docstring to build the
    # tool schema it sends the model. There is no hand-written schema to keep
    # in sync — the function is the tool.
    print(f"[agent] search_threads(keyword_filter={keyword_filter!r})")
    return search.run_search(keyword_filter, limit=3)


def get_thread(thread_id: int) -> Dict[str, Any]:
    """Get the full summary of one Hacker News discussion thread.

    Use this when the user asks about a specific discussion you have already
    found — for example "what was the second one about?" or "tell me more about
    that Rust thread". Search results carry only a title and a one-line description;
    this returns what was actually discussed, plus the link to it.

    Args:
        thread_id: The id of the thread, taken from an earlier search result.
    """
    print(f"[agent] get_thread(thread_id={thread_id!r})")
    thread = corpus.get(thread_id)
    if thread is None:
        return {
            "error": (
                "No thread with that id is in the last 30 days of data. "
                "Search for the topic instead."
            )
        }
    return corpus.public_view(thread, detail=True)


_agent = LlmAgent(
    model=LiteLlm(
        model="openai/gemma",
        api_base=f"{LLAMA_SERVER_URL}/v1",
        api_key="not-needed",
        # Gemma bills chain-of-thought against max_tokens and will spend the
        # whole budget thinking, returning empty content. Disabled the same way
        # llm.chat_completion does — see the note there.
        extra_body={"chat_template_kwargs": {"enable_thinking": False}},
    ),
    name=APP_NAME,
    instruction=BASE_INSTRUCTION,
    tools=[search_threads, get_thread],
)

_session_service: DatabaseSessionService | None = None
_runner: Runner | None = None


def start() -> None:
    """Open the session database and build the runner.

    Sessions persist rather than living in memory: reopening the app and asking
    a follow-up on an old conversation has to still work, and an in-memory store
    loses that on every restart.
    """
    global _session_service, _runner
    os.makedirs(DB_DIR, exist_ok=True)
    _session_service = DatabaseSessionService(db_url=f"sqlite+aiosqlite:///{DB_PATH}")
    _runner = Runner(agent=_agent, app_name=APP_NAME, session_service=_session_service)
    print(f"[agent] sessions at {DB_PATH}")


async def send(
    session_id: str,
    message: str,
    pinned_thread: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Run one turn.

    session_id is the frontend's conversation id, so a Dexie conversation and an
    ADK session are the same thing seen from two sides — the UI store holds
    rendered prose, the agent store holds the structure it reasons over.

    Returns the reply, the tool calls made, and the raw results those tools
    produced. The frontend renders links from `results` rather than from the
    model's prose: the model retypes a 50-character URL every time it writes a
    list, and one wrong character is a dead link nobody notices until it's
    clicked.
    """
    if _runner is None or _session_service is None:
        raise RuntimeError("agent not started — call start() first")

    session = await _session_service.get_session(
        app_name=APP_NAME, user_id=USER_ID, session_id=session_id
    )
    if session is None:
        session = await _session_service.create_session(
            app_name=APP_NAME, user_id=USER_ID, session_id=session_id
        )

    print(f"[agent] attached={'yes: ' + str(pinned_thread.get('story_title')) if pinned_thread else 'none'}")

    reply_parts: list[str] = []
    tool_calls: list[Dict[str, Any]] = []
    results: list[Dict[str, Any]] = []

    async for event in _runner.run_async(
        user_id=USER_ID,
        session_id=session.id,
        new_message=types.Content(
            role="user",
            parts=[types.Part(text=_with_attachment(message, pinned_thread))],
        ),
    ):
        if not (event.content and event.content.parts):
            continue
        for part in event.content.parts:
            call = getattr(part, "function_call", None)
            response = getattr(part, "function_response", None)
            if call:
                tool_calls.append({"name": call.name, "arguments": dict(call.args or {})})
            elif response:
                # Keep the structured output so the UI can render it directly.
                # search_threads returns {"results": [...]}; get_thread returns a
                # single thread. Both need capturing, or a turn that reads one
                # discussion leaves the user with prose and no way to open it.
                payload = response.response
                if not isinstance(payload, dict):
                    continue
                if isinstance(payload.get("results"), list):
                    results.extend(payload["results"])
                elif payload.get("thread_id"):
                    # The link comes from the corpus, not from the payload the
                    # model saw — see corpus.public_view. Looking it up here is
                    # also what guarantees it is right: a model retyping a
                    # 50-character URL gets one character wrong eventually, and
                    # a dead link is only noticed when somebody clicks it.
                    row = corpus.get(payload["thread_id"])
                    if row:
                        results.append({
                            "thread_id": row.get("thread_id"),
                            "title": row.get("story_title"),
                            "link": row.get("anchor"),
                        })
            elif getattr(part, "text", None):
                reply_parts.append(part.text)

    # The model emits text at several points in the loop — a preamble before a
    # tool call, then the real answer. The last non-empty part is the answer;
    # the earlier ones are it narrating what it is about to fetch.
    reply = next((t.strip() for t in reversed(reply_parts) if t.strip()), "")
    return {"reply": reply, "tool_calls": tool_calls, "results": results}

File: src-tauri/src/agent.rs
//! The agent, in Tauri core.
//!
//! Replaces `sidecar/agent.py`. The model still runs in llama-server on :8081 —
//! only the agent framework moved, ADK to Rig. The corpus and search still live
//! in the Python sidecar, so the tools below reach them over HTTP the way the
//! frontend already does; #48 removes that hop.
//!
//! The problem the agent solves is unchanged: the model calls a tool, sees the
//! result, and decides again. Routing used to give it one tool call per turn and
//! then stop, so when a question needed something the model didn't have it
//! invented it instead. Every hallucination this project hit came from that gap.

use std::sync::{Arc, Mutex};

use rig_agent::agent::AgentBuilder;
use rig_agent::prelude::*;
use rig_agent::tool::{Tool, ToolContext};
use rig_core::providers::llamafile;
use serde::{Deserialize, Serialize};
use serde_json::json;

/// Where the Python sidecar still serves the corpus. Removed by #48.
const SIDECAR_URL: &str = "http://localhost:8000";
/// The inference server. Replaced by in-process llama-cpp-2 in #47.
const LLAMA_SERVER_URL: &str = "http://localhost:8081";

/// Carried over verbatim from agent.py. The "or links" clause is load-bearing:
/// without it the model builds `item?id={thread_id}` out of the id it was given
/// and writes it into its prose, and the UI then renders the same link a second
/// time. Observed while spiking this port.
const BASE_INSTRUCTION: &str = concat!(
    "You help a user explore Hacker News. Use search_threads to find ",
    "discussions on a topic, and get_thread to read one in detail when the ",
    "user asks about a specific result. Answer from what the tools return — do ",
    "not invent discussions, titles, or links. If you do not have enough ",
    "information to answer, say so plainly rather than guessing. Do not show ",
    "thread ids to the user; they are for your own tool calls.",
    "\n\nA message may begin with a note about a discussion the user has open ",
    "in front of them. When they say \"this thread\", \"this discussion\", ",
    "\"this\", or \"this one\", they mean that one — not an item from a list ",
    "of search results. Answer from the note; there is no need to search."
);

/// The thread the user has dragged into the chat, as the frontend sends it.
#[derive(Debug, Clone, Deserialize)]
pub struct PinnedThread {
    pub story_title: Option<String>,
    pub theme: Option<String>,
    pub summary: Option<String>,
}

/// The pinned thread travels WITH the message it was sent alongside, rather
/// than living in agent state.
///
/// State was the wrong home for it, and failed in a way worth recording. State
/// feeds the system prompt, which has no position in time — as far as the model
/// can tell it has always said this — while a thread just read with get_thread
/// sits in the newest message. Asked to explain "this thread", the model picked
/// the recent one. No instruction wording outranked it.
fn with_attachment(message: &str, pinned: Option<&PinnedThread>) -> String {
    match pinned {
        None => message.to_string(),
        Some(t) => format!(
            "[The user has this discussion open in front of them:\n\"{}\" — {}.\nWhat people said: {}]\n\n{}",
            // It's t.story_title or ""
            t.story_title.as_deref().unwrap_or(""),
            t.theme.as_deref().unwrap_or(""),
            t.summary.as_deref().unwrap_or(""),
            message
        ),
    }
}

/// One thread as the UI renders it. `link` never reaches the model — see the
/// note on GetThread.
#[derive(Debug, Clone, Serialize)]
pub struct UiResult {
    pub thread_id: u64,
    pub title: String,
    pub link: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct ToolCallRecord {
    pub name: String,
    pub arguments: serde_json::Value,
}

/// What the tools report back for the UI, filled in as the agent loop runs.
///
/// The frontend renders links from `results` rather than from the model's prose:
/// the model retypes a 50-character URL every time it writes a list, and one
/// wrong character is a dead link nobody notices until it's clicked.
#[derive(Default)]
struct Collected {
    tool_calls: Vec<ToolCallRecord>,
    results: Vec<UiResult>,
}

type Shared = Arc<Mutex<Collected>>;

#[derive(Debug, thiserror::Error)]
#[error("{0}")]
pub struct ToolFailed(String);

// ---------------------------------------------------------------------------
// search_threads
// ---------------------------------------------------------------------------

#[derive(Debug, Deserialize)]
pub struct SearchArgs {
    keyword_filter: String,
}

struct SearchThreads {
    http: reqwest::Client,
    collected: Shared,
}

impl Tool for SearchThreads {
    const NAME: &'static str = "search_threads";
    type Args = SearchArgs;
    type Output = serde_json::Value;
    type Error = ToolFailed;

    fn description(&self) -> String {
        "Search Hacker News discussion threads by topic.".into()
    }

    fn parameters(&self) -> serde_json::Value {
        json!({
            "type": "object",
            "properties": {
                "keyword_filter": {
                    "type": "string",
                    "description": "The topic or keywords to search discussions for."
                }
            },
            "required": ["keyword_filter"]
        })
    }

    async fn call(
        &self,
        _ctx: &mut ToolContext,
        args: Self::Args,
    ) -> Result<Self::Output, Self::Error> {
        self.collected.lock().unwrap().tool_calls.push(ToolCallRecord {
            name: Self::NAME.into(),
            arguments: json!({ "keyword_filter": args.keyword_filter }),
        });

        let body = self
            .http
            .post(format!("{SIDECAR_URL}/search"))
            .json(&json!({ "query": args.keyword_filter, "limit": 3 }))
            .send()
            .await
            .map_err(|e| ToolFailed(e.to_string()))?
            .json::<serde_json::Value>()
            .await
            .map_err(|e| ToolFailed(e.to_string()))?;

        // Search lists options; get_thread is how you explore one. A hit carries
        // neither the summary nor a link — that is what stops the UI appending a
        // second copy of the list the model just wrote.
        Ok(body)
    }
}

// ---------------------------------------------------------------------------
// get_thread
// ---------------------------------------------------------------------------

#[derive(Debug, Deserialize)]
pub struct ThreadArgs {
    thread_id: u64,
}

struct GetThread {
    http: reqwest::Client,
    collected: Shared,
}

impl Tool for GetThread {
    const NAME: &'static str = "get_thread";
    type Args = ThreadArgs;
    type Output = serde_json::Value;
    type Error = ToolFailed;

    fn description(&self) -> String {
        "Get the full summary of one Hacker News discussion thread. Use this when \
         the user asks about a specific discussion you have already found — for \
         example \"what was the second one about?\". Search results carry only a \
         title and a one-line description; this returns what was actually discussed."
            .into()
    }

    fn parameters(&self) -> serde_json::Value {
        json!({
            "type": "object",
            "properties": {
                "thread_id": {
                    "type": "integer",
                    "description": "The id of the thread, taken from an earlier search result."
                }
            },
            "required": ["thread_id"]
        })
    }

    async fn call(
        &self,
        _ctx: &mut ToolContext,
        args: Self::Args,
    ) -> Result<Self::Output, Self::Error> {
        self.collected.lock().unwrap().tool_calls.push(ToolCallRecord {
            name: Self::NAME.into(),
            arguments: json!({ "thread_id": args.thread_id }),
        });

        let mut body = self
            .http
            .get(format!("{SIDECAR_URL}/thread/{}", args.thread_id))
            .send()
            .await
            .map_err(|e| ToolFailed(e.to_string()))?
            .json::<serde_json::Value>()
            .await
            .map_err(|e| ToolFailed(e.to_string()))?;

        // The link is split off here and never handed to the model. Given a URL
        // it writes it into its prose, and the UI renders the same link again
        // from `results`. Taking it out is what makes that impossible rather
        // than merely discouraged.
        if let Some(link) = body.get("link").and_then(|v| v.as_str()).map(str::to_string) {
            let title = body
                .get("title")
                .and_then(|v| v.as_str())
                .unwrap_or_default()
                .to_string();
            self.collected.lock().unwrap().results.push(UiResult {
                thread_id: args.thread_id,
                title,
                link,
            });
        }
        if let Some(obj) = body.as_object_mut() {
            obj.remove("link");
        }

        Ok(body)
    }
}

// ---------------------------------------------------------------------------
// The command the frontend calls
// ---------------------------------------------------------------------------

/// The durable store Rig loads from and appends to. Built once at startup.
pub struct AgentMemory(pub Arc<crate::memory::SqliteConversationMemory>);

#[derive(Debug, Default, Serialize)]
pub struct AgentReply {
    pub reply: String,
    pub tool_calls: Vec<ToolCallRecord>,
    pub results: Vec<UiResult>,
    /// Failures come back as a normal reply with this set, not as a rejected
    /// promise: the frontend already renders `response.error` inline, and a
    /// throw would bypass that and surface as an unhandled rejection instead.
    pub error: Option<String>,
}

impl AgentReply {
    fn failed(message: impl Into<String>) -> Self {
        Self {
            error: Some(message.into()),
            ..Default::default()
        }
    }
}

/// Gemma takes a minute or two to load 12GB, and until it has, every request
/// fails with a bare connection error. Saying so plainly is worth a round trip:
/// a not-yet-ready server has twice sent us chasing bugs that did not exist.
async fn llama_server_ready(http: &reqwest::Client) -> bool {
    matches!(
        http.get(format!("{LLAMA_SERVER_URL}/health"))
            .timeout(std::time::Duration::from_secs(5))
            .send()
            .await,
        Ok(response) if response.status().is_success()
    )
}

#[tauri::command]
pub async fn send_message(
    memory: tauri::State<'_, AgentMemory>,
    session_id: String,
    message: String,
    pinned_thread: Option<PinnedThread>,
) -> Result<AgentReply, String> {
    Ok(run_turn(memory.0.clone(), session_id, message, pinned_thread).await)
}

/// One turn, with no Tauri in sight, so the check runner can drive the real
/// agent rather than a copy of it that might drift.
pub async fn run_turn(
    memory: Arc<crate::memory::SqliteConversationMemory>,
    session_id: String,
    message: String,
    pinned_thread: Option<PinnedThread>,
) -> AgentReply {
    let collected: Shared = Arc::new(Mutex::new(Collected::default()));
    let http = reqwest::Client::new();

    if !llama_server_ready(&http).await {
        return AgentReply::failed("The model server is still starting up.");
    }

    println!(
        "[agent] attached={}",
        match pinned_thread.as_ref().and_then(|t| t.story_title.as_deref()) {
            Some(title) => format!("yes: {title}"),
            None => "none".to_string(),
        }
    );

    let client = match llamafile::Client::from_url(LLAMA_SERVER_URL) {
        Ok(client) => client,
        Err(error) => return AgentReply::failed(error.to_string()),
    };
    let model = client.completion_model("gemma");

    let agent = AgentBuilder::new(model)
        .preamble(BASE_INSTRUCTION)
        .tool(SearchThreads {
            http: http.clone(),
            collected: collected.clone(),
        })
        .tool(GetThread {
            http,
            collected: collected.clone(),
        })
        // Rig runs ONE model call by default, so without this any tool use dies
        // with MaxTurnsError before the model ever answers. ADK looped by default.
        .default_max_turns(10)
        // Gemma bills chain-of-thought against the same max_tokens as the answer
        // and will spend all of it thinking, returning empty content with
        // finish_reason=Length.
        .additional_params(json!({
            "chat_template_kwargs": { "enable_thinking": false }
        }))
        .temperature(0.1)
        .max_tokens(2048)
        // Rig loads this conversation's history before the turn and appends the
        // new messages after it. Doing it here rather than by hand is what makes
        // the store swappable: the agent code never touches SQLite.
        .memory(memory)
        .conversation(session_id)
        .build();

    let prompt = with_attachment(&message, pinned_thread.as_ref());
    let reply = match agent.prompt(prompt).await {
        Ok(reply) => reply,
        Err(error) => {
            println!("[agent] failed: {error}");
            return AgentReply::failed(error.to_string());
        }
    };

    let out = collected.lock().unwrap();
    // turns ≈ tool calls + 1. Logged so the default_max_turns cap can be set
    // from real usage rather than guessed at.
    println!("[agent] tool calls: {}", out.tool_calls.len());
    AgentReply {
        reply,
        tool_calls: out.tool_calls.clone(),
        results: out.results.clone(),
        error: None,
    }
}
